python/controller_selector.py

```diff
@@ -0,0 +1,268 @@
+
+        int : current_time
+            The current time (unix timestamp)
+
+        int : latest_switch_time
+            The time of the latest switch of the controller mode
+
+        float : heading_error
+            The current heading error in radians
+
+        """
+        tooSlow = boat_speed <= sailbot_constants.SPEED_THRESHOLD_FOR_JIBING_KNOTS
+        timeout = current_time - latest_switch_time >= sailbot_constants.MAX_TIME_FOR_TACKING
+        reached_heading = abs(heading_error) <= sailbot_constants.MIN_HEADING_ERROR_FOR_SWITCH
+
+        # TODO: Panic mode. If we are not at the desired heading by the end of the timeout then
+        #       we need to do something else.
+        if(tooSlow or timeout or reached_heading):
+            self.__assignMode(ControlModes.UNKNOWN.value)
+        return
+
+    def __switchFromJibing(self, current_time, latest_switch_time, heading_error):
+        """
+        The controller switches from JIBE_ONLY mode to UNKNOWN mode if it meets its
+        exit condition. The boat must be in JIBE_ONLY mode when this method is invoked
+        or there may be unexpected results.
+
+        Exit Condition
+        --------------
+        - The boat has been in JIBE_ONLY mode for longer than the specified timeout, OR
+        - The boat has reached the specified setpoint
+
+        Arguments
+        ---------
+        int : current_time
+            The current time (unix timestamp)
+
+        int : latest_switch_time
+            The time of the latest switch of the controller mode
+
+        float : heading_error
+            The current heading error in radians
+
+        """
+        timeout = current_time - latest_switch_time >= sailbot_constants.MAX_TIME_FOR_JIBING
+        reached_heading = abs(heading_error) <= sailbot_constants.MIN_HEADING_ERROR_FOR_SWITCH
+
+        if(timeout or reached_heading):
+            self.__assignMode(ControlModes.UNKNOWN.value)
+        return
+
+    def __assignMode(self, control_mode_id):
+        """
+        Changes the control mode according to the control mode ID. Refer to
+        control_modes.py for the control mode IDs and how they are encoded.
+        If an invalid control mode ID is passed, then the default control mode
+        is tacking, but the control mode ID is assigned to UNKNOWN. Note that an
+        invalid control mode ID is not ideal and should be avoided.
+
+        Arguments
+        ---------
+        control_mode_id : int
+            The ID of the new control mode. Refer to control_mode.py for the possible
+            control mode IDs and what control modes they map to.
+
+        """
+
+        # Control mode is valid
+        if(self.isValidModeID(control_mode_id)):
+            self.__controlModeID = control_mode_id
+            self.__controlMode = self.controllerMappings.get(self.__controlModeID, TackController)
+
+        # Invalid control mode passed. Assigning a default value
+        else:
```

Comment:
just curious why are we not raising exception in this case?

Reply:
To be honest, I only raise an exception if the user input is bad. We should actually never get to this point in the code if the user input is correct and everything else internally is correct. I figured that we would not want to raise an exception if something failed internally since that would be bad for the boat, so I figured that staying in TACKABLE mode would be a fallback option without stopping execution.